File: apps/server/models/workflow.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime


class Position(BaseModel):
    x: float
    y: float


class EventFilter(BaseModel):
    event: str
    condition: Optional[str] = None


class NodeModel(BaseModel):
    id: str
    type: str
    position: Position
    config: Dict[str, Any] = Field(default_factory=dict)
    event_filters: Optional[List[EventFilter]] = Field(default=None, alias="eventFilters")

    class Config:
        populate_by_name = True


class ConnectionEndpoint(BaseModel):
    node_id: str = Field(alias="nodeId")
    port: str

    class Config:
        populate_by_name = True


class ConnectionModel(BaseModel):
    id: str
    from_endpoint: ConnectionEndpoint = Field(alias="from")
    to_endpoint: ConnectionEndpoint = Field(alias="to")

    class Config:
        populate_by_name = True


class CharacterConfig(BaseModel):
    name: str = "AI Assistant"
    personality: str = "Friendly and helpful virtual streamer"


class WorkflowCreate(BaseModel):
    name: str
    description: Optional[str] = None
    nodes: List[NodeModel] = Field(default_factory=list)
    connections: List[ConnectionModel] = Field(default_factory=list)
    character: CharacterConfig = Field(default_factory=CharacterConfig)


class WorkflowUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    nodes: Optional[List[NodeModel]] = None
    connections: Optional[List[ConnectionModel]] = None
    character: Optional[CharacterConfig] = None


class WorkflowResponse(BaseModel):
    id: str
    name: str
    description: Optional[str] = None
    nodes: List[NodeModel] = Field(default_factory=list)
    connections: List[ConnectionModel] = Field(default_factory=list)
    character: CharacterConfig
    created_at: datetime = Field(alias="createdAt")
    updated_at: datetime = Field(alias="updatedAt")

    class Config:
        populate_by_name = True
        from_attributes = True


class ExecutionStatus(BaseModel):
    workflow_id: str
    status: str  # "idle", "running", "stopped", "error"
    started_at: Optional[datetime] = None
    error: Optional[str] = None


class ExecutionRequest(BaseModel):
    """Request body for starting workflow execution with current state."""
    nodes: Optional[List[NodeModel]] = None
    connections: Optional[List[ConnectionModel]] = None
    character: Optional[CharacterConfig] = None

File: apps/server-ts/src/models/workflow.ts
import { z } from "zod";

// ─── Primitives ─────────────────────────────────────────────────

export const positionSchema = z.object({
  x: z.number(),
  y: z.number(),
});

export const eventFilterSchema = z.object({
  event: z.string(),
  condition: z.string().optional(),
});

// ─── Node & Connection ──────────────────────────────────────────

export const nodeModelSchema = z.object({
  id: z.string(),
  type: z.string(),
  position: positionSchema,
  config: z.record(z.any()).default({}),
  eventFilters: z.array(eventFilterSchema).optional(),
});

export const connectionEndpointSchema = z.object({
  nodeId: z.string(),
  port: z.string(),
});

export const connectionModelSchema = z.object({
  id: z.string(),
  from: connectionEndpointSchema,
  to: connectionEndpointSchema,
});

// ─── Character ──────────────────────────────────────────────────

export const characterConfigSchema = z.object({
  name: z.string().default("AI Assistant"),
  personality: z.string().default("Friendly and helpful virtual streamer"),
});

// ─── Workflow CRUD ──────────────────────────────────────────────

export const workflowCreateSchema = z.object({
  name: z.string(),
  description: z.string().optional(),
  nodes: z.array(nodeModelSchema).default([]),
  connections: z.array(connectionModelSchema).default([]),
  character: characterConfigSchema.default({}),
});

export const workflowUpdateSchema = z.object({
  name: z.string().optional(),
  description: z.string().optional(),
  nodes: z.array(nodeModelSchema).optional(),
  connections: z.array(connectionModelSchema).optional(),
  character: characterConfigSchema.optional(),
});

export const workflowResponseSchema = z.object({
  id: z.string(),
  name: z.string(),
  description: z.string().optional(),
  nodes: z.array(nodeModelSchema).default([]),
  connections: z.array(connectionModelSchema).default([]),
  character: characterConfigSchema,
  createdAt: z.string(),
  updatedAt: z.string(),
});

// ─── Execution ──────────────────────────────────────────────────

export const executionStatusSchema = z.object({
  workflowId: z.string(),
  status: z.enum(["idle", "running", "stopped", "error"]),
  startedAt: z.string().optional(),
  error: z.string().optional(),
});

export const executionRequestSchema = z.object({
  nodes: z.array(nodeModelSchema).optional(),
  connections: z.array(connectionModelSchema).optional(),
  character: characterConfigSchema.optional(),
  startNodeId: z.string().optional(),
});

// ─── Inferred Types ─────────────────────────────────────────────

export type Position = z.infer<typeof positionSchema>;
export type EventFilter = z.infer<typeof eventFilterSchema>;
export type NodeModel = z.infer<typeof nodeModelSchema>;
export type ConnectionEndpoint = z.infer<typeof connectionEndpointSchema>;
export type ConnectionModel = z.infer<typeof connectionModelSchema>;
export type CharacterConfig = z.infer<typeof characterConfigSchema>;
export type WorkflowCreate = z.infer<typeof workflowCreateSchema>;
export type WorkflowUpdate = z.infer<typeof workflowUpdateSchema>;
export type WorkflowResponse = z.infer<typeof workflowResponseSchema>;
export type ExecutionStatus = z.infer<typeof executionStatusSchema>;
export type ExecutionRequest = z.infer<typeof executionRequestSchema>;
